pkg/actuators/awsplacementgroup/controller.go

```diff
@@ -3,91 +3,622 @@ package awsplacementgroup
 	}
 
 	awsPlacementGroup.Status = *originalAWSPlacementGroupStatus
 
 	if err := r.Client.Status().Patch(ctx, awsPlacementGroup, originalAWSPlacementGroupToPatch); err != nil {
-		return ctrl.Result{}, fmt.Errorf("failed to patch AWSPlacementGroup Status: %v", err)
+		return ctrl.Result{}, fmt.Errorf("failed to patch AWSPlacementGroup Status: %w", err)
 	}
 
 	return result, err
 }
 
-func (r *Reconciler) reconcile(ctx context.Context, logger logr.Logger, awsPlacementGroup *machinev1.AWSPlacementGroup) (ctrl.Result, error) {
-	logger.Info("reconciling AWSPlacementGroup")
+// reconcile reconciles an AWSPlacementGroup.
+func (r *AWSPlacementGroupReconciler) reconcile(ctx context.Context, logger logr.Logger, awsPlacementGroup *machinev1.AWSPlacementGroup) (ctrl.Result, error) {
+	now := metav1.Now()
+	if awsPlacementGroup.Status.ExpiresAt == nil ||
+		awsPlacementGroup.Status.ExpiresAt.Before(&now) {
+		// The cached ObservedConfiguration stored in the AWSPlacementGroup Status
+		// is expired or not present. Proceed with the syncing.
+		// Check AWS for the configuration of the placement group and reflect this in the status of the object.
+		if err := reflectObservedConfiguration(r.AWSClient, logger, awsPlacementGroup); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	// Update Status.ManagementState with the observed spec value.
+	awsPlacementGroup.Status.ManagementState = awsPlacementGroup.Spec.ManagementSpec.ManagementState
+
+	// Only proceed if the AWSPlacementGroup is Managed.
+	if awsPlacementGroup.Spec.ManagementSpec.ManagementState == machinev1.UnmanagedManagementState {
+		// This AWSPlacementGroup is now Unmanaged so clean up any machine finalizer if there is any
+		// as placement group resources on AWS shouldn't be deleted if they are running in Unmanaged mode.
+		if controllerutil.ContainsFinalizer(awsPlacementGroup, awsPlacementGroupFinalizer) {
+			controllerutil.RemoveFinalizer(awsPlacementGroup, awsPlacementGroupFinalizer)
+			logger.Info("removing finalizer from AWSPlacementGroup")
+			// Return without requeuing.
+			// The after all the finalizers have been removed the object should be
+			// cleaned up by the garbage collector without the need of a requeue.
+			return ctrl.Result{}, nil
+		}
+
+		logger.Info("ignoring Unmanaged AWSPlacementGroup")
+		// Return and requeue the placement group even if it is Unmanaged to keep syncing up its ObservedConfiguration.
+		return ctrl.Result{RequeueAfter: requeueIn(awsPlacementGroup.Status.ExpiresAt.Time)}, nil
+	}
+
+	// If object DeletionTimestamp is zero, it means the object is not being deleted
+	// so clean up relevant resources.
+	if awsPlacementGroup.DeletionTimestamp.IsZero() {
+		if !controllerutil.ContainsFinalizer(awsPlacementGroup, awsPlacementGroupFinalizer) {
+			controllerutil.AddFinalizer(awsPlacementGroup, awsPlacementGroupFinalizer)
+			logger.Info("adding finalizer to AWSPlacementGroup")
+			// Return to update the object with it.
+			return ctrl.Result{}, nil
```

Comment:
Perhaps requeue true if we are expecting to re-reconcile imediately?

Reply:
Yep, good catch!